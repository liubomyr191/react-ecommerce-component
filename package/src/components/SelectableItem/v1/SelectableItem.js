import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import uniqueId from "lodash.uniqueid";
import { withComponents } from "@reactioncommerce/components-context";
import { addTypographyStyles, applyTheme } from "../../../utils";

const StyledLabel = styled.label`
  ${addTypographyStyles("SelectableItemLabel", "labelText")}
  align-items: center;
  cursor: pointer;
  display: flex;
`;

const StyledRadioButton = styled.span`
  align-items: center;
  background-color: ${applyTheme("selectableItemRadioButtonBackgroundColor")};
  border-radius: 50%;
  border-color: ${applyTheme("selectableItemRadioButtonBorderColor")};
  border-style: solid;
  border-width: ${applyTheme("selectableItemRadioButtonBorderWidth")};
  box-sizing: border-box;
  display: flex;
  flex-shrink: 0;
  height: ${applyTheme("selectableItemRadioButtonSize")};
  justify-content: center;
  margin-bottom: 0;
  margin-left: 0;
  margin-right: ${applyTheme("selectableItemRadioButtonSpacingToLabel")};
  margin-top: 0;
  width: ${applyTheme("selectableItemRadioButtonSize")};
`;

const StyledInput = styled.input`
  border: 0;
  clip: rect(0, 0, 0, 0);
  cursor: pointer;
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  visibility: visible;
  white-space: nowrap;
  width: 1px;
  &:checked + ${StyledLabel} ${StyledRadioButton}::before {
    content: " ";
    display: inline-block;
    position: relative;
    width: ${applyTheme("selectableItemRadioButtonCheckSize")};
    height: ${applyTheme("selectableItemRadioButtonCheckSize")};
    border-radius: 50%;
    background-color: ${applyTheme("selectableItemRadioButtonColor")};
  }
  &:focus + ${StyledLabel} ${StyledRadioButton} {
    box-shadow: ${applyTheme("selectableItemRadioFocus")};
    outline: ${applyTheme("selectableItemRadioFocusOutline")}
  }
  &:disabled + ${StyledLabel} ${StyledRadioButton} {
    background-color: ${applyTheme("selectableItemRadioDisabledFillColor")};
  }
  &:disabled + ${StyledLabel} {
    opacity: ${applyTheme("selectableItemRadioDisabledOpacity")};
  }
  &:disabled + ${StyledLabel}:hover {
    cursor:  ${applyTheme("selectableItemRadioDisabledCursor")};
  }
`;

const StyledDetail = styled.div`
  ${addTypographyStyles("SelectableItemDetail", "bodyText")}
  align-items: center;
  display: flex;
  justify-content: center;
`;

const StyledIcon = styled.span`
  border-radius: ${applyTheme("selectableListBorderRadius")};
  border-color: ${applyTheme("selectableListBorderColor")};
  border-style: ${applyTheme("selectableListBorderStyle")};
  border-width: ${applyTheme("selectableListBorderWidth")};
  height: ${applyTheme("selectableListIconHeight")};
  margin-bottom: 0;
  margin-left: 0;
  margin-right: ${applyTheme("selectableListIconSpacingToLabel")};
  margin-top: 0;
  width: ${applyTheme("selectableListIconWidth")};
  svg {
    height: ${applyTheme("selectableListIconHeight")};
    width: ${applyTheme("selectableListIconWidth")};
  }
`;

const StyledItem = styled.div`
  display: flex;
  justify-content: space-between;
  height: ${applyTheme("selectableListHeight")};
  @media (max-width: 768px) {
    height: ${applyTheme("selectableListHeightMobile")};
  }
`;

const LeftAlignedItem = styled.div`
  display: flex;
  justify-content: flex-start;
  height: ${applyTheme("selectableListHeight")};
  @media (max-width: 768px) {
    height: ${applyTheme("selectableListHeightMobile")};
  }
  ${StyledLabel} {
    position: relative;
    font-weight: ${applyTheme("selectableListLeftAlignedLabelFontWeight")};
  }
  ${StyledDetail} {
    ${addTypographyStyles("SelectableItemDetail", "labelText")}
    margin-left: ${applyTheme("selectableListLeftAlignedDetailSpacingToLabel")};
  }
`;


class SelectableItem extends Component {
  static propTypes = {
    /**
     * You can provide a `className` prop that will be applied to the outermost DOM element
     * rendered by this component. We do not recommend using this for styling purposes, but
     * it can be useful as a selector in some situations.
     */
    className: PropTypes.string,
    /**
     * Optional text, SVG or element displayed on the right-hand side
     */
    detail: PropTypes.node,
    /**
     * Optional icon (SVG) displayed on the left-hand side
     */
    icon: PropTypes.node,
    /**
     * `true` if the item is checked
     */
    isChecked: PropTypes.bool,
    /**
     * Left-aligned style
     */
    isLeftAligned: PropTypes.bool,
    /**
     * Read only and disabled state
     */
    isReadOnly: PropTypes.bool,
    /**
     * Label
     */
    label: PropTypes.string.isRequired,
    /**
     * Called whenever this item becomes selected or unselected. Two arguments are provided,
     * `isChecked` and the `value` prop that was passed in.
     */
    onChange: PropTypes.func,
    /**
     * A value to be passed to `onChange`
     */
    value: PropTypes.string.isRequired
  }

  static defaultProps = {
    onChange() { },
    isChecked: false,
    isLeftAligned: false,
    isReadOnly: false
  };

  uniqueInstanceIdentifier = uniqueId("SelectableItem_");

  handleChange = (event) => {
    const { value } = this.props;
    this.props.onChange(event.target.checked, value);
  };

  render() {
    const {
      className,
      detail,
      icon,
      isChecked,
      isLeftAligned,
      isReadOnly,
      label,
      value
    } = this.props;

    const id = `radio_${this.uniqueInstanceIdentifier}`;

    const input = (
      <StyledInput
        id={id}
        checked={isChecked}
        value={value}
        onChange={this.handleChange}
        type="radio"
        disabled={isReadOnly}
      />
    );

    const labelAndButton = (
      <StyledLabel
        htmlFor={id}
      >
        <StyledRadioButton />
        {icon ? <StyledIcon>{icon}</StyledIcon> : null}
        {label}
      </StyledLabel>
    );

    return (
      <div className={className}>
        {isLeftAligned ?
          <LeftAlignedItem>
            {input}
            {labelAndButton}
            {detail ? <StyledDetail>{detail}</StyledDetail> : null}
          </LeftAlignedItem >
          :
          <StyledItem>
            {input}
            {labelAndButton}
            {detail ? <StyledDetail>{detail}</StyledDetail> : null}
          </StyledItem >
        }
      </div>
    );
  }
}

export default withComponents(SelectableItem);
