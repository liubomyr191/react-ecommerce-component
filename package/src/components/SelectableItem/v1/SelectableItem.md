### Overview

`SelectableItems` in a `SelectableList` are used to select shipping methods, addresses and credit cards. Each item consists of a radio button, label and display value. The label and display value can both support text or render other elements, like icons and links.

An individual item can be checked, or unchecked, along with disabled (`isReadOnly`) or enabled:

```jsx noeditor
const item = { _id: "1", label: "Default unchecked" };
<SelectableItem item={item} />
```

```jsx noeditor
const item = { _id: "123", label: "Default checked" };
<SelectableItem item={item} value={true} />
```

```jsx noeditor
const item = { _id: "12", label: "Default disabled and unchecked" };
<SelectableItem item={item} value={false} isReadOnly />
```

```jsx noeditor
const item = { _id: "12", label: "Default disabled and checked" };
<SelectableItem item={item} value={true} isReadOnly />
```

#### Usage
- All labels should be written in sentence caps.
- Radio buttons should be used when only one option can be selected from a set of defined values.

#### Types

##### SelectableItem without `detail`

```jsx
const item = { _id: "2", label: "Default address" };
<SelectableItem item={item} />
```

##### SelectableItem with `detail`

Pass any element - text, SVGs or React elements - into `detail` to display a secondary action on the right-hand side.

###### Plain text

```jsx
const item = {_id: "3", label: "Free shipping", detail: "$0.00"};
<SelectableItem item={item}/>
```

```jsx
const item = {_id: "4", label: "Free shipping", detail: "\u2714"};
<SelectableItem item={item}/>
```

###### Element

```jsx
const link = (
    <Button title="Default" className="myBtn" isTextOnly isShortHeight>Default Text</Button>
);
const item = {_id: "5", label: "Default", detail: link};

<SelectableItem item={item}/>
```

###### SVG
```jsx
const iconClear = (
  // credit: https://fontawesome.com/icons/times-circle?style=regular
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 512 512"
    style={{ height: "20px", maxHeight: "20px", verticalAlign: "middle" }}
  >
    <path
      fill="#3c3c3c"
      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
    />
  </svg>
);
const item = {_id: "6", label: "Free shipping", detail: iconClear};
<SelectableItem item={item}/>
```


#### Specs

|Property                                |Style                                |
|----------------------------------------|:-----------------------------------:|
|Label font                              | `@rui-label-text` capitalized       |
|Label color                             | `@cool-grey-500`                    |
|Border color                            | `@cool-grey-500`                    |
|Radio button color                      | `@cool-grey-500`                    |
|Radio button height and width           | 20px                                |
|Radio button and label spacing          | 11px                                |
|Radio button border                     | 2px solid `@cool-grey-500`          |
|Selected circle size                    | 11px                                |
|Cursor                                  | Pointer (radio button and label)    |