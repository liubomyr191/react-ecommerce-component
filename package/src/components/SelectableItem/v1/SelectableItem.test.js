import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import SelectableItem from "./SelectableItem";

test("basic snapshot", () => {
  const component = renderer.create(<SelectableItem />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
