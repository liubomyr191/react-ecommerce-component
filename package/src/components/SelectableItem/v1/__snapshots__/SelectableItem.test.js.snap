// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot with empty props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 50px;
}

.c0 input {
  cursor: pointer;
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  border: 0;
  visibility: visible;
  white-space: nowrap;
}

.c0 input:checked + label .radio::before {
  content: " ";
  display: inline-block;
  position: relative;
  width: 10px;
  height: 10px;
  border-radius: 50%;
  background-color: #3c3c3c;
}

.c0 input:focus + label .radio {
  box-shadow: 0 0 0 2px #8ce0c9;
  outline: 1px solid transparent;
}

.c0 input:disabled + label .radio {
  background-color: #e6e6e6;
}

.c0 input:disabled + label {
  opacity: .5;
}

.c0 input:disabled + label:hover {
  cursor: not-allowed;
}

.c0 label {
  font-family: 'Source Sans Pro','Helvetica Neue',Helvetica,sans-serif;
  color: #3c3c3c;
  font-size: 14px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .radio {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #ffffff;
  border: 2px solid #3c3c3c;
  height: 20px;
  width: 20px;
  margin: 0 10px 0 0;
  border-radius: 50%;
  box-sizing: border-box;
}

.c0 .icon {
  border-radius: 2px;
  border: 1px solid #e6e6e6;
  margin: 0 10px 0 0;
  width: 38px;
  height: 24px;
}

.c0 .icon svg {
  width: 38px;
  height: 24px;
}

.c0 .detail {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-family: 'Source Sans Pro','Helvetica Neue',Helvetica,sans-serif;
  font-size: 16px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
}

@media (max-width:768px) {
  .c0 {
    height: 60px;
  }
}

<div
  className="c0"
>
  <input
    checked={false}
    disabled={undefined}
    id={undefined}
    name={undefined}
    onChange={[Function]}
    type="radio"
  />
  <label
    htmlFor={undefined}
  >
    <span
      className="radio"
    />
    Label
  </label>
  <div
    className="detail"
  >
    Detail
  </div>
</div>
`;
